### PR TITLE
k8s.io/apiserver: remove skewed completion from EtcdOptions

### PR DESCRIPTION
**Pull request by sttts**
Rule of thumb:
- options = flags
- configs = things making up a process, but not running yet and not only wired where necessary
- server = a process, everything wired, can be run.

Rule of thumb number 2: no backwards dependencies

Here, EtcdOptions depended on a config of another delegated apiserver, plus it made the EtcdOptions "run" something. This is a skewed dependency, making composition of components hard.

This PR moves the EtcdOptions->config dependency by storing the resource transformers in the config. It's not a 100% solution as also a config should not "run". Something for a follow-up. But that's less painful and is rather cosmetics.

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by enj** [comment]
Unit test failure looks real: `etcd_test.go:313: healthz checks are not equal, missing=[], extra=["kms-provider-0" "kms-provider-1"]`

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 1]
```diff
@@ -306,24 +242,99 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
...
+	}
+
+	ctxServer := wait.ContextForChannel(c.DrainedNotify())
+	ctxTransformers, closeTransformers := context.WithCancel(ctxServer)
+	defer func() {
+		// in case of error, we want to close partially initialized (if any) transformers
+		if err != nil {
+			closeTransformers()
+		}
+	}()
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:294]: I forget the rules for named returns, does this always run, even if the error is shadow assigned in an inner scope?

**Comment by sttts** [comment]
> `pull-kubernetes-dependencies` is failing because `k8s.io/controller-manager` seems to have pulled in `k8s.io/kms` somehow (which does not sound correct).

This is correct. kube-controller-manager is an apiserver based on k8s.io/apiserver (for metrics and other stuff), like all of our components. We now pulled your encryption code into the generic apiserver config. Before one had to pull in the options.

How bad is that dependency? Do we care? If yes, we can generalize the resource transformer interface by moving it out of encryptionconfig. Then the dependency should go away. Either way is fine.

**Comment by enj** [comment]
> How bad is that dependency? Do we care? If yes, we can generalize the resource transformer interface by moving it out of encryptionconfig. Then the dependency should go away.

I think it would be simpler to just make a copy of the resource transformer interface in the API server code (without moving it out of encryptionconfig) so it no longer has the direct dependency.  The KMS repo isn't that large but I don't see why would increase the KCM dep tree to include it, especially with an easy and straightforward solution.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 2]
```diff
@@ -297,7 +228,12 @@ func (s *EtcdOptions) ApplyTo(c *server.Config) error {
 		return nil
 	}
 
-	return s.ApplyWithStorageFactoryTo(&SimpleStorageFactory{StorageConfig: s.StorageConfig}, c)
+	storageConfig := s.StorageConfig
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:231]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:294]: it does

**Comment by sttts** [comment]
@enj addressed everything.

**Comment by sttts** [comment]
Infra flake.

/retest

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 3]
```diff
@@ -212,118 +208,123 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
...
-		return nil
+func (s *EtcdOptions) CreateSimpleRESTOptionsGetter(resourceTransformers storagevalue.ResourceTransformers, tracker flowcontrolrequest.StorageObjectCountTracker) generic.RESTOptionsGetter {
+	storageConfigCopy := s.StorageConfig
+	if storageConfigCopy.StorageObjectCountTracker == nil {
+		storageConfigCopy.StorageObjectCountTracker = tracker
 	}
 
-	return s.ApplyWithStorageFactoryTo(&SimpleStorageFactory{StorageConfig: s.StorageConfig}, c)
+	return s.CreateRESTOptionsGetter(&SimpleStorageFactory{StorageConfig: s.StorageConfig}, resourceTransformers)
 }
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:253]: IMO this logic should just be inlined into `NewCRDRESTOptionsGetter` (it is only used there).

**Review by enj — CHANGES_REQUESTED** [review]
Mostly minor stuff, I would probably just squash this into one commit.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go [hunk 4]
```diff
@@ -129,30 +126,16 @@ func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, err
...
-// We assume that the input etcd options have been completed already.
+//
 // Avoid messing with anything outside of changes to StorageConfig as that
 // may lead to unexpected behavior when the options are applied.
-func NewCRDRESTOptionsGetter(etcdOptions genericoptions.EtcdOptions) (genericregistry.RESTOptionsGetter, error) {
+func NewCRDRESTOptionsGetter(etcdOptions genericoptions.EtcdOptions, resourceTransformers storagevalue.ResourceTransformers, tracker flowcontrolrequest.StorageObjectCountTracker) genericregistry.RESTOptionsGetter {
+	// operate on a copy, don't mutate the original
 	etcdOptions.StorageConfig.Codec = unstructured.UnstructuredJSONScheme
 	etcdOptions.WatchCacheSizes = nil      // this control is not provided for custom resources
 	etcdOptions.SkipHealthEndpoints = true // avoid double wiring of health checks
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:136]: done

File: cmd/kube-apiserver/app/apiextensions.go [hunk 5]
```diff
@@ -32,6 +32,7 @@ import (
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/apiserver/pkg/util/webhook"
 	kubeexternalinformers "k8s.io/client-go/informers"
+
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-apiserver/app/apiextensions.go:35]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 6]
```diff
@@ -212,118 +208,123 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
...
+			delegate:             factory,
+			resourceTransformers: resourceTransformers,
+		}
 	}
+	return &StorageFactoryRestOptionsFactory{Options: *s, StorageFactory: factory}
+}
 
-	if !s.complete {
-		return fmt.Errorf("EtcdOptions.Apply called without completion")
+func (s *EtcdOptions) applyResourceTransformers(c *server.Config) (err error) {
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:265]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:253]: yep, looks better. Done.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go [hunk 7]
```diff
@@ -129,30 +126,16 @@ func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, err
...
-	restOptionsGetter := c.RESTOptionsGetter
-	if restOptionsGetter == nil {
-		return nil, fmt.Errorf("server.Config RESTOptionsGetter should not be nil")
-	}
-	// sanity check that no other fields are set
-	c.RESTOptionsGetter = nil
-	if !reflect.DeepEqual(c, genericapiserver.Config{}) {
-		return nil, fmt.Errorf("only RESTOptionsGetter should have been mutated in server.Config")
-	}
-	return restOptionsGetter, nil
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:155]: Not sure I can follow. This path is explicitly non-standard for CRs, and it is simple, calling the options package. So no risk to diverge.

What is still a little strange IMO is that the tracker is not set on the StorageConfig persistently. But that's another skew for another PR.

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:155]: If the `Apply` funcs add further logic beyond what is covered by the tracker/transformer/etc, then CRs would diverge from standard resources again (in the same way they did for transformers which broke encryption for them).  This is a slow moving part of the codebase so maybe it is fine.

**Comment by enj** [comment]
/lgtm
/hold
/assign @benluddy
(David requested that Ben review this PR as well)

**Review by benluddy — COMMENTED** [review]
/lgtm
/hold cancel

**Pull request merged**